vslib/src/sai_vs_hostintf.cpp

```diff
@@ -29,7 +67,414 @@ sai_status_t vs_send_hostif_packet(
     return SAI_STATUS_NOT_IMPLEMENTED;
 }
 
-VS_GENERIC_QUAD(HOSTIF,hostif);
+int vs_create_tap_device(const char *dev, int flags)
+{
+    SWSS_LOG_ENTER();
+
+    const char *tundev = "/dev/net/tun";
+
+    int fd = open(tundev, O_RDWR);
+
+    if (fd < 0)
+    {
+        SWSS_LOG_ERROR("failed to open %s", tundev);
+
+        return -1;
+    }
+
+    struct ifreq ifr;
+
+    memset(&ifr, 0, sizeof(ifr));
+
+    ifr.ifr_flags = (short int)flags;  // IFF_TUN or IFF_TAP, IFF_NO_PI
+
+    strncpy(ifr.ifr_name, dev, IFNAMSIZ);
+
+    int err = ioctl(fd, TUNSETIFF, (void *) &ifr);
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("ioctl TUNSETIFF on fd %d %s failed, err %d", fd, dev, err);
+
+        close(fd);
+
+        return err;
+    }
+
+    return fd;
+}
+
+int vs_set_dev_mac_address(const char *dev, const sai_mac_t mac)
+{
+    SWSS_LOG_ENTER();
+
+    int s = socket(AF_INET, SOCK_DGRAM, 0);
+
+    if (s < 0)
+    {
+        SWSS_LOG_ERROR("failed to create socket, errno: %d", errno);
+
+        return -1;
+    }
+
+    struct ifreq ifr;
+
+    strncpy(ifr.ifr_name, dev, IFNAMSIZ);
+
+    memcpy(ifr.ifr_hwaddr.sa_data, mac, 6);
+
+    ifr.ifr_hwaddr.sa_family = ARPHRD_ETHER;
+
+    int err = ioctl(s, SIOCSIFHWADDR, &ifr);
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("ioctl SIOCSIFHWADDR on socket %d %s failed, err %d", s, dev, err);
+    }
+
+    close(s);
+
+    return err;
+}
+
+int ifup(const char *dev)
+{
+    int s = socket(AF_INET, SOCK_DGRAM, 0);
+
+    if (s < 0)
+    {
+        SWSS_LOG_ERROR("failed to open socket: %d", s);
+
+        return -1;
+    }
+
+    struct ifreq ifr;
+
+    memset(&ifr, 0, sizeof ifr);
+
+    strncpy(ifr.ifr_name, dev , IFNAMSIZ);
+
+    ifr.ifr_flags |= IFF_UP;
+
+    int err = ioctl(s, SIOCSIFFLAGS, &ifr);
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("ioctl SIOCSIFFLAGS on socket %d %s failed, err %d", s, dev, err);
+    }
+
+    close(s);
+
+    return err;
+}
+
+void veth2tap_fun(std::shared_ptr<hostif_info_t> info)
+{
+    SWSS_LOG_ENTER();
+
+    unsigned char buffer[0x4000];
+
+    while (info->run_thread)
+    {
+        // TODO convert to non blocking using select
+
+        ssize_t size = read(info->packet_socket, buffer, sizeof(buffer));
+
+        if (size < 0)
+        {
+            SWSS_LOG_ERROR("failed to read from socket %d", info->packet_socket);
+            break;
+        }
+
+        // TODO examine packet for mac and possible generate fdb_event
+
+        if (write(info->tapfd, buffer, size) < 0)
+        {
+            SWSS_LOG_ERROR("failed to write to tap device %d", info->tapfd);
+            break;
+        }
+    }
+}
+
+void tap2veth_fun(std::shared_ptr<hostif_info_t> info)
+{
+    SWSS_LOG_ENTER();
+
+    unsigned char buffer[0x4000];
```

Comment:
Just a note - 16K bytes allocated from stack.

Reply:
Also in above function.

---
In reply to: [144614375](https://github.com/Azure/sonic-sairedis/pull/244#discussion_r144614375) [](ancestors = 144614375)

Reply:
should be fine, the thread stack size is 8MB for 64-bit os and 2MB for 32-bit os

---
In reply to: [144627332](https://github.com/Azure/sonic-sairedis/pull/244#discussion_r144627332) [](ancestors = 144627332,144614375)

Reply:
shouldbe fine